ergo-wallet/src/main/scala/org/ergoplatform/wallet/boxes/ErgoBoxAssetExtractor.scala

```diff
@@ -0,0 +1,89 @@
+package org.ergoplatform.wallet.boxes
+
+import akka.util.ByteString
+import org.ergoplatform.ErgoBoxCandidate
+import sigmastate.eval.Extensions._
+
+import scala.collection.mutable
+import scala.util.Try
+
+object ErgoBoxAssetExtractor {
+  val MaxAssetsPerBox = 255
+
+  /**
+    * Extracts a mapping (assets -> total amount) from a set of boxes passed as a parameter.
+    * That is, the method is checking amounts of assets in the boxes(i.e. that a box contains positive
+    * amount for an asset) and then summarize and group their corresponding amounts.
+    *
+    * @param boxes - boxes to check and extract assets from
+    * @return a mapping from asset id to balance and total assets number
+    */
+  def extractAssets(
+    boxes: IndexedSeq[ErgoBoxCandidate]
+  ): Try[(Map[ByteString, Long], Int)] = Try {
+    val map: mutable.Map[ByteString, Long] = mutable.Map[ByteString, Long]()
+    val assetsNum = boxes.foldLeft(0) {
+      case (acc, box) =>
+        require(
+          box.additionalTokens.length <= ErgoBoxAssetExtractor.MaxAssetsPerBox,
+          "too many assets in one box"
+        )
+        box.additionalTokens.foreach {
+          case (assetId, amount) =>
+            val aiWrapped = ByteString.fromArray(assetId)
+            val total     = map.getOrElse(aiWrapped, 0L)
+            map.put(aiWrapped, Math.addExact(total, amount))
+        }
+        acc + box.additionalTokens.size
+    }
+    map.toMap -> assetsNum
+  }
+
+  /**
+    * Cost of assets preservation rules checks.
+    * We iterate through all assets to create a map (cost: `(outAssetsNum + inAssetsNum) * tokenAccessCost)`)
+    * and after that we iterate through unique asset ids to check preservation rules
+    *     (cost: `(inAssets.size + outAssets.size) * tokenAccessCost`)
+    * @param inAssetsNum number of input assets in all boxes
+    * @param inAssetsSize number if unique input asset ids
+    * @param outAssetsNum number of output assets in all boxes
+    * @param outAssetsSize number if unique output asset ids
+    * @param tokenAccessCost access cost for a token
+    * @return total assets access cost
+    */
+  def totalAssetsAccessCost(
+    inAssetsNum: Int,
+    inAssetsSize: Int,
+    outAssetsNum: Int,
+    outAssetsSize: Int,
+    tokenAccessCost: Int
+  ): Int =
+    (outAssetsNum + inAssetsNum) * tokenAccessCost + (inAssetsSize + outAssetsSize) * tokenAccessCost
```

Comment:
Cost accumulation should always throw on arithmetic overflow.

Reply:
@pragmaxim please use addExact / multiplyExact here

Reply:
please use it from `java7.compat.Math` class since this is in wallet module and is used in appkit.